### PR TITLE
planner: enable column pruning for common handle

### LINKED ISSUE #22642: some sql plans with clustered index are worse the none clustered index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Initialization:
[prepare_sql.txt](https://github.com/pingcap/tidb/files/5902349/prepare_sql.txt)
```sql
create database with_cluster_index;
create database wout_cluster_index;
use with_cluster_index;
set @@tidb_enable_clustered_index = 1;
source prepare_sql.txt;
analyze table tbl_0;analyze table tbl_1;analyze table tbl_2;analyze table tbl_3;analyze table tbl_4;
use wout_cluster_index;
set @@tidb_enable_clustered_index = 0;
source prepare_sql.txt;
analyze table tbl_0;analyze table tbl_1;analyze table tbl_2;analyze table tbl_3;analyze table tbl_4;
```

query:
```sql
select count(*) from tbl_0 where col_0 < 5429  ;
```

### 2. What did you expect to see? (Required)
The plan of clustered_index and none_clustered_index are the same.

### 3. What did you see instead (Required)
```sql
select count(*) from tbl_0 where col_0 < 5429  ;
Performance with cluster index: 
avg:2302.719040, sum:230271.904000, p80:2499.922000, p90:2773.655000, p95:2958.653000, max:3861.342000, min:1864.088000
Performance wout cluster index: 
avg:976.619680, sum:97661.968000, p80:1091.711000, p90:1488.636000, p95:2348.649000, max:3381.523000, min:673.036000
with_cluster_index_plan: 
StreamAgg_27	1.00	1	root		time:1.68ms, loops:2	funcs:count(Column#9)->Column#6	N/A	N/A
└─TableReader_28	1.00	1	root		time:1.68ms, loops:2, cop_task: {num: 1, max: 1.64ms, proc_keys: 2244, tot_proc: 4ms, rpc_num: 1, rpc_time: 1.63ms, copr_cache: disabled}	data:StreamAgg_9	227 Bytes	N/A
  └─StreamAgg_9	1.00	1	cop[tikv]		tikv_task:{time:4ms, loops:3}, scan_detail: {total_process_keys: 2244, total_keys: 2245, rocksdb: {delete_skipped_count: 2305, key_skipped_count: 4549, block: {cache_hit_count: 9, read_count: 0, read_byte: 0 Bytes}}}	funcs:count(1)->Column#9	N/A	N/A
    └─Selection_26	798.90	799	cop[tikv]		tikv_task:{time:4ms, loops:3}, scan_detail: {total_process_keys: 0, total_keys: 0, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 0, read_count: 0, read_byte: 0 Bytes}}}	lt(with_cluster_index.tbl_0.col_0, 5429)	N/A	N/A
      └─TableFullScan_25	2244.00	2244	cop[tikv]	table:tbl_0	tikv_task:{time:4ms, loops:3}, scan_detail: {total_process_keys: 0, total_keys: 0, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 0, read_count: 0, read_byte: 0 Bytes}}}	keep order:false	N/A	N/A
wout_cluster_index_plan: 
StreamAgg_17	1.00	1	root		time:502.3µs, loops:2	funcs:count(Column#9)->Column#7	N/A	N/A
└─IndexReader_18	1.00	1	root		time:499.7µs, loops:2, cop_task: {num: 1, max: 469.8µs, proc_keys: 799, rpc_num: 1, rpc_time: 459.3µs, copr_cache: disabled}	index:StreamAgg_9	213 Bytes	N/A
  └─StreamAgg_9	1.00	1	cop[tikv]		tikv_task:{time:0s, loops:1}, scan_detail: {total_process_keys: 799, total_keys: 800, rocksdb: {delete_skipped_count: 269, key_skipped_count: 1068, block: {cache_hit_count: 3, read_count: 0, read_byte: 0 Bytes}}}	funcs:count(1)->Column#9	N/A	N/A
    └─IndexRangeScan_16	798.90	799	cop[tikv]	table:tbl_0, index:idx_3(col_0)	tikv_task:{time:0s, loops:1}, scan_detail: {total_process_keys: 0, total_keys: 0, rocksdb: {delete_skipped_count: 0, key_skipped_count: 0, block: {cache_hit_count: 0, read_count: 0, read_byte: 0 Bytes}}}	range:[-inf,5429), keep order:false	N/A	N/A
```

### 4. What is your TiDB version? (Required)

nightly

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #22642

Problem Summary:

In https://github.com/pingcap/tidb/pull/18391, we skipped common handles when doing the column pruning for simplicity, this caused some unnecessary table read when reading index is enough.

### What is changed and how it works?

1. enable column pruning for common handles (see changes in rule_column_pruning.go)
2. add common handles to tablescan's schema when it's the double read case (find_best_task.go)
3. do not add a projection above PhysicalIndexLookupReader if we did the agg pushdown (task.go)
4. also add common handles to tablescan's schema in the index merge join when it's the double read case (exhaust_physical_plans.go)

Note: In this PR, all columns in the common handle are always read in the table worker of index lookup reader no matter whether it requires to keep order or not. This can be improved after resolving https://github.com/pingcap/tidb/issues/23037

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- no release note.

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -1032,6 +1032,15 @@ func (ds *DataSource) convertToIndexScan(prop *property.PhysicalProperty, candid
 	task = cop
 	if cop.tablePlan != nil && ds.tableInfo.IsCommonHandle {
 		cop.commonHandleCols = ds.commonHandleCols
+		commonHandle := ds.handleCols.(*CommonHandleCols)
+		for _, col := range commonHandle.columns {
```
Comment (lzmhhh123) [inline planner/core/find_best_task.go:1036]: Would this logic also be implemented in `func (p *LogicalJoin) constructInnerIndexScanTask`?

Comment (time-and-fate) [reply planner/core/find_best_task.go:1036]: Yes. I constructed a SQL which would fail to run without this logic.

Comment (time-and-fate) [reply planner/core/find_best_task.go:1036]: added this logic into `constructInnerIndexScanTask` and added the case into unit tests.

**Review by lzmhhh123 — APPROVED** [review]

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -1032,6 +1032,15 @@ func (ds *DataSource) convertToIndexScan(prop *property.PhysicalProperty, candid
...
+			if ds.schema.ColumnIndex(col) == -1 {
+				ts := cop.tablePlan.(*PhysicalTableScan)
+				ts.Schema().Append(col)
+				ts.Columns = append(ts.Columns, col.ToInfo())
+				cop.doubleReadNeedProj = true
```
Comment (winoros) [inline planner/core/find_best_task.go:1041]: We can just break here?

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -1022,6 +1022,16 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 			physicalTableID: ds.physicalTableID,
 		}.Init(ds.ctx, ds.blockOffset)
 		ts.schema = is.dataSourceSchema.Clone()
+		if ds.tableInfo.IsCommonHandle {
+			commonHandle := ds.handleCols.(*CommonHandleCols)
+			for _, col := range commonHandle.columns {
+				if ts.schema.ColumnIndex(col) == -1 {
+					ts.Schema().Append(col)
+					ts.Columns = append(ts.Columns, col.ToInfo())
+					cop.doubleReadNeedProj = true
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go:1031]: break here directly too?

File: planner/core/find_best_task.go (hunk 2 above)
Comment (time-and-fate) [reply planner/core/find_best_task.go:1041]: We need all the columns in the common handle.

Comment (winoros) [reply planner/core/find_best_task.go:1041]: ok, got it

**Review by winoros — COMMENTED** [review]
lgtm

**Review by winoros — APPROVED** [review]

**Pull request merged**